### PR TITLE
feature: report gpu resource

### LINKED ISSUE #381: [proposal] koordlet collects basic devices info and create/update Device CRD
**What is your proposal**:
koordlet needs to create or update Device CRD instance with the resources information and healthy status per device including GPU, RDMA and FPGA, etc. This version we only support GPU.

**Why is this needed**:
koord-scheduler needs info in this CRD to make scheduling decisions.

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: Yue Zhang <huaihuan.zy@alibaba-inc.com>

### Ⅱ. Does this pull request fix one issue?
fixes #381

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -40,6 +40,7 @@ require (
 	github.com/Azure/go-ansiterm v0.0.0-20210617225240-d185dfc1b5a1 // indirect
 	github.com/Microsoft/go-winio v0.4.15 // indirect
 	github.com/Microsoft/hcsshim v0.8.10-0.20200715222032-5eafd1556990 // indirect
+	github.com/NVIDIA/gpu-monitoring-tools v0.0.0-20211102125545-5a2c58442e48
```
Comment (jasonliu747) [inline go.mod:43]: it's deprecated, use `github.com/NVIDIA/go-nvml` instead.

**Review by jasonliu747 — CHANGES_REQUESTED** [review]

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func (s *statesInformer) reportGPUDevice() {
+	s.nodeRWMutex.RLock()
+	nodeName := s.node.Name
+	s.nodeRWMutex.RUnlock()
+	queryParam := generateQueryParam()
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
+	var deviceInfos []schedulingv1alpha1.DeviceInfo
+	for _, gpu := range nodeResource.Metric.GPUs {
```
Comment (LambdaHJ) [inline pkg/koordlet/statesinformer/states_gpu.go:37]: check nodeResource.Error

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+			Type:   schedulingv1alpha1.GPU,
+			Health: health,
+			Resources: map[corev1.ResourceName]resource.Quantity{
+				extension.GPUCore:        *resource.NewQuantity(100, resource.BinarySI),
+				extension.GPUMemory:      gpu.MemoryTotal,
+				extension.GPUMemoryRatio: *resource.NewQuantity(100, resource.BinarySI),
+			},
+		})
+	}
+	if len(deviceInfos) == 0 {
```
Comment (LambdaHJ) [inline pkg/koordlet/statesinformer/states_gpu.go:55]: Check it on line 35, len(nodeResource.Metric.GPUs) == 0

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	device := &schedulingv1alpha1.Device{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+		},
+		Spec: schedulingv1alpha1.DeviceSpec{
+			Devices: deviceInfos,
+		},
+	}
+	_, err := s.schedulingClient.Devices().Get(context.TODO(), nodeName, metav1.GetOptions{})
```
Comment (LambdaHJ) [inline pkg/koordlet/statesinformer/states_gpu.go:67]: Call Update directly.

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		return
+	}
+}
+
+func (s *statesInformer) GPUHealCheck() {
+	// FIXME: metrics cache ready?
+	queryParam := generateQueryParam()
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
+	var deviceInfos []schedulingv1alpha1.DeviceInfo
+	for _, gpu := range nodeResource.Metric.GPUs {
```
Comment (LambdaHJ) [inline pkg/koordlet/statesinformer/states_gpu.go:91]: nodeResource.Error

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func (s *statesInformer) reportGPUDevice() {
+	s.nodeRWMutex.RLock()
+	nodeName := s.node.Name
+	s.nodeRWMutex.RUnlock()
+	queryParam := generateQueryParam()
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
+	if len(nodeResource.Metric.GPUs) == 0 {
+		klog.Infof("no gpu device found")
```
Comment (jasonliu747) [inline pkg/koordlet/statesinformer/states_gpu.go:54]: please set the correct log level according to our convention, thanks.

Comment (jasonliu747) [inline pkg/koordlet/statesinformer/states_gpu.go:54]: please set the correct log level according to our convention, thanks.

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		End:       &end,
+	}
+}
+
+func (s *statesInformer) reportGPUDevice() {
+	s.nodeRWMutex.RLock()
+	nodeName := s.node.Name
+	s.nodeRWMutex.RUnlock()
+	queryParam := generateQueryParam()
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
```
Comment (jasonliu747) [inline pkg/koordlet/statesinformer/states_gpu.go:52]: you probably need to check `nodeResource.Error != nil` as well

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+	device := &schedulingv1alpha1.Device{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+		},
+		Spec: schedulingv1alpha1.DeviceSpec{
+			Devices: deviceInfos,
+		},
+	}
+	_, err := s.deviceClient.Update(context.TODO(), device, metav1.UpdateOptions{})
```
Comment (hormes) [inline pkg/koordlet/statesinformer/states_gpu.go:89]: 先比较一下是否有变化，避免频繁去写 apiserver。实现上，为了保险期间，可以做成这样
if 有变化 or 上次汇报时间 > xx {
  update
}

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/features"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+)
+
+func generateQueryParam() *metriccache.QueryParam {
+	end := time.Now()
+	start := end.Add(-time.Duration(60) * time.Second)
+	return &metriccache.QueryParam{
+		Aggregate: metriccache.AggregationTypeAVG,
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_gpu.go:41]: 只上报拓扑信息，用latest查询一下就可以了，不需要聚合

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+	nodeName := s.node.Name
+	s.nodeRWMutex.RUnlock()
+	queryParam := generateQueryParam()
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
+	if nodeResource.Error != nil {
+		klog.Errorf("failed to get node resource metric, err: %v", nodeResource.Error)
+		return
+	}
+	if len(nodeResource.Metric.GPUs) == 0 {
+		klog.V(4).Info("no gpu device found")
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_gpu.go:58]: 日志会刷屏，降到V(6)

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+			Resources: map[corev1.ResourceName]resource.Quantity{
+				extension.GPUCore:        *resource.NewQuantity(100, resource.BinarySI),
+				extension.GPUMemory:      gpu.MemoryTotal,
+				extension.GPUMemoryRatio: *resource.NewQuantity(100, resource.BinarySI),
+			},
+		})
+	}
+
+	device := &schedulingv1alpha1.Device{
+		ObjectMeta: metav1.ObjectMeta{
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_gpu.go:83]: 补充一个owner ref，node下线时自动清理

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+		_, err = s.deviceClient.Create(context.TODO(), device, metav1.CreateOptions{})
+		if err != nil {
+			klog.Errorf("failed to create device %s, err: %v", nodeName, err)
+			return
+		}
+	}
+}
+
+func (s *statesInformer) GPUHealCheck() {
+	if !features.DefaultKoordletFeatureGate.Enabled(features.Accelerators) {
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_gpu.go:105]: feature gate移动到最外层，没开启的话不启动任何线程

File: pkg/slo-controller/noderesource/noderesource_controller.go [hunk 13]
```diff
@@ -104,11 +107,21 @@ func (r *NodeResourceReconciler) Reconcile(ctx context.Context, req ctrl.Request
 
 	beResource := r.calculateBEResource(node, podList, nodeMetric)
 
-	if err := r.updateNodeBEResource(node, beResource); err != nil {
+	if err := r.updateBENodeResource(node, beResource); err != nil {
 		klog.Errorf("failed to update node %v BE resource, error: %v", node.Name, err)
 		return ctrl.Result{Requeue: true}, err
 	}
 
+	device, err := r.DeviceClient.Get(context.TODO(), node.Name, metav1.GetOptions{})
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/noderesource_controller.go:115]: 没开启或者非gpu的节点，应该忽略更新，不然这里会刷屏。

File: pkg/slo-controller/noderesource/noderesource.go [hunk 14]
```diff
@@ -112,10 +113,49 @@ func (r *NodeResourceReconciler) resetNodeBEResource(node *corev1.Node, reason,
...
-	return r.updateNodeBEResource(node, beResource)
+	return r.updateBENodeResource(node, beResource)
 }
 
-func (r *NodeResourceReconciler) updateNodeBEResource(node *corev1.Node, beResource *nodeBEResource) error {
+func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device *schedulingv1alpha1.Device) error {
+	if device == nil {
+		return nil
+	}
+	memoryTotal := resource.NewQuantity(0, resource.BinarySI)
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/noderesource.go:123]: 确认一下这几个的单位，感觉只有Memory的应该是BinarySI

File: pkg/slo-controller/noderesource/noderesource.go [hunk 15]
```diff
@@ -112,10 +113,49 @@ func (r *NodeResourceReconciler) resetNodeBEResource(node *corev1.Node, reason,
...
+	coreTotal := resource.NewQuantity(0, resource.BinarySI)
+	ratioTotal := resource.NewQuantity(0, resource.BinarySI)
+	for _, gpu := range device.Spec.Devices {
+		if gpu.Health {
+			memoryTotal.Add(gpu.Resources[extension.GPUMemory])
+			coreTotal.Add(gpu.Resources[extension.GPUCore])
+			ratioTotal.Add(gpu.Resources[extension.GPUMemoryRatio])
+		}
+	}
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/noderesource.go:133]: 非必要不更新，可以单独加一个Device的watch event

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 10 above)
Comment (jasonliu747) [reply pkg/koordlet/statesinformer/states_gpu.go:58]: v5 would be the highest

File: pkg/slo-controller/noderesource/noderesource.go [hunk 16]
```diff
@@ -112,10 +113,49 @@ func (r *NodeResourceReconciler) resetNodeBEResource(node *corev1.Node, reason,
...
+
+		if err := r.Client.Status().Update(context.TODO(), updateNode); err != nil {
+			klog.Errorf("failed to update node gpu resource %v, error: %v", updateNode.Name, err)
+			return err
+		}
+		return nil
+	})
+}
+
+func (r *NodeResourceReconciler) updateBENodeResource(node *corev1.Node, beResource *nodeBEResource) error {
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/noderesource.go:158]: we might need to figure out a new function name

File: pkg/slo-controller/noderesource/noderesource.go [hunk 17]
```diff
@@ -112,10 +113,49 @@ func (r *NodeResourceReconciler) resetNodeBEResource(node *corev1.Node, reason,
...
+			}
+			return err
+		}
+
+		updateNode.Status.Capacity[extension.GPUMemory] = *memoryTotal
+		updateNode.Status.Allocatable[extension.GPUMemory] = *memoryTotal
+		updateNode.Status.Capacity[extension.GPUCore] = *coreTotal
+		updateNode.Status.Allocatable[extension.GPUCore] = *coreTotal
+		updateNode.Status.Capacity[extension.GPUMemoryRatio] = *ratioTotal
+		updateNode.Status.Allocatable[extension.GPUMemoryRatio] = *ratioTotal
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/noderesource.go:148]: Capacity我理解就是满的，但Allocatable应该是会时时变化？

Comment (jasonliu747) [reply pkg/slo-controller/noderesource/noderesource.go:148]: 我可能是错的：https://github.com/NVIDIA/k8s-device-plugin/issues/213

File: pkg/slo-controller/noderesource/noderesource.go (hunk 16 above)
Comment (ZYecho) [reply pkg/slo-controller/noderesource/noderesource.go:158]: why?

File: pkg/slo-controller/noderesource/noderesource_controller.go (hunk 13 above)
Comment (ZYecho) [reply pkg/slo-controller/noderesource/noderesource_controller.go:115]: ok

File: pkg/slo-controller/noderesource/noderesource.go (hunk 16 above)
Comment (jasonliu747) [reply pkg/slo-controller/noderesource/noderesource.go:158]: 我理解不止是BE resource了？

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 8 above)
Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_gpu.go:100]: 这个地方没有做informer，要做diff的话，需要先拿到oldDevice，需要多一次get，所以留了一个TODO，后边支持

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+			health = false
+		}
+		s.gpuMutex.RUnlock()
+		deviceInfos = append(deviceInfos, schedulingv1alpha1.DeviceInfo{
+			UUID:   gpu.DeviceUUID,
+			Minor:  gpu.Minor,
+			Type:   schedulingv1alpha1.GPU,
+			Health: health,
+			Resources: map[corev1.ResourceName]resource.Quantity{
+				extension.GPUCore:        *resource.NewQuantity(100, resource.BinarySI),
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_gpu.go:74]: DecimalSI？

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 8 above)
Comment (zwzhang0107) [reply pkg/koordlet/statesinformer/states_gpu.go:100]: 和上次cpu topo crd的更新一起，建一个issue记录一下吧

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+			continue
+		}
+		uuid, ret := gpudevice.GetUUID()
+		if ret != nvml.SUCCESS {
+			klog.Errorf("failed to get device uuid at index %d, err: %v", deviceIndex, nvml.ErrorString(ret))
+		}
+		devices = append(devices, uuid)
+	}
+	unhealthyChan := make(chan string)
+	go checkHealth(context.TODO(), devices, unhealthyChan)
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_gpu.go:138]: 把stopCh传进来替换掉context.TODO()吧

File: pkg/slo-controller/noderesource/device_event_handler.go [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+)
+
+var _ handler.EventHandler = &EnqueueRequestForDevice{}
+
+type EnqueueRequestForDevice struct {
+	client.Client
+	syncContext *SyncContext
+}
+
+func (n *EnqueueRequestForDevice) Create(e event.CreateEvent, q workqueue.RateLimitingInterface) {
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/device_event_handler.go:41]: 创建时也需要q.Add，尤其是后面非必要不更新后

File: pkg/slo-controller/noderesource/nodemetric_event_handler.go [hunk 21]
```diff
@@ -78,7 +77,8 @@ func (n *EnqueueRequestForNodeMetric) cleanSyncContext(nodeMetric *slov1alpha1.N
 		return fmt.Errorf("failed to cleanup empty sync context")
 	}
 
-	n.syncContext.Delete(util.GetNodeMetricKey(nodeMetric))
+	// nodeMetric's name = node's name
+	n.syncContext.Delete(fmt.Sprintf("%v/%v", nodeMetric.GetNamespace(), nodeMetric.GetName()))
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/nodemetric_event_handler.go:81]: GetNodeKey(ObjectMeta)

**Comment by jasonliu747** [comment]
@caohe you might be interested in checking this PR

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+	nodeResource := s.metricsCache.GetNodeResourceMetric(queryParam)
+	if nodeResource.Error != nil {
+		klog.Errorf("failed to get node resource metric, err: %v", nodeResource.Error)
+		return
+	}
+	if len(nodeResource.Metric.GPUs) == 0 {
+		klog.V(5).Info("no gpu device found")
+		return
+	}
+	var deviceInfos []schedulingv1alpha1.DeviceInfo
```
Comment (buptcozy) [inline pkg/koordlet/statesinformer/states_gpu.go:60]: consider for fpga\rdma's extensibility in the future,  we should not bind the gpu device crd generation process and update\create logic together.   I think the below process is better:

1. var deviceInfos []schedulingv1alpha1.DeviceInfo
2. fill gpu\fpga\rdma to deviceInfos
3. create\update remote crd

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+		}
+		devices = append(devices, uuid)
+	}
+	unhealthyChan := make(chan string)
+	go checkHealth(stopCh, devices, unhealthyChan)
+	klog.Info("start to do gpu health check")
+	for d := range unhealthyChan {
+		// FIXME: there is no way to recover from the Unhealthy state.
+		s.gpuMutex.Lock()
+		s.unhealthyGPU[d] = struct{}{}
```
Comment (buptcozy) [inline pkg/koordlet/statesinformer/states_gpu.go:143]: I don't find move gpu minor out of unhealthyGPU logic if card turns good.

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+		s.gpuMutex.Unlock()
+		klog.Infof("get a unhealthy gpu %s", d)
+	}
+}
+
+// check status of gpus, and send unhealthy devices to the unhealthyDeviceChan channel
+func checkHealth(stopCh <-chan struct{}, devs []string, xids chan<- string) {
+	eventSet, ret := nvml.EventSetCreate()
```
Comment (buptcozy) [inline pkg/koordlet/statesinformer/states_gpu.go:151]: I notice that the society apply a config for disable HealthCheck,  may be for user to apply its self health check logic, should we follow?

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 22 above)
Comment (buptcozy) [inline pkg/koordlet/statesinformer/states_gpu.go:60]: If we add rdma\fpga in future,  the function structure should be:

1. var deviceInfos []schedulingv1alpha1.DeviceInfo
2. add gpu\fpga\rdma to deviceInfos
3. update\create crd

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+	end := time.Now()
+	start := end.Add(-time.Duration(60) * time.Second)
+	return &metriccache.QueryParam{
+		Aggregate: metriccache.AggregationTypeLast,
+		Start:     &start,
+		End:       &end,
+	}
+}
+
+func (s *statesInformer) reportGPUDevice() {
```
Comment (buptcozy) [inline pkg/koordlet/statesinformer/states_gpu.go:46]: also need update deviceStatus of crd.

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 23 above)
Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_gpu.go:143]: from the original code of asi device plugin，there's a comment

// FIXME: there is no way to recover from the Unhealthy state.

please fixme if error

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 24 above)
Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_gpu.go:151]: sorry, can't get you point

File: pkg/koordlet/statesinformer/states_gpu.go [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+		}
+		_, err = s.deviceClient.Create(context.TODO(), device, metav1.CreateOptions{})
+		if err != nil {
+			klog.Errorf("failed to create device %s, err: %v", copyNode.Name, err)
+			return
+		}
+	}
+}
+
+func (s *statesInformer) GPUHealCheck(stopCh <-chan struct{}) {
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_gpu.go:114]: fixed

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 22 above)
Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_gpu.go]: thanks, refactor it

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 25 above)
Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_gpu.go]: later if need it

File: pkg/koordlet/statesinformer/states_device.go [hunk 27]
```diff
@@ -0,0 +1,212 @@
...
+				extension.GPUMemory:      gpu.MemoryTotal,
+				extension.GPUMemoryRatio: *resource.NewQuantity(100, resource.DecimalSI),
+			},
+		})
+	}
+	return deviceInfos
+}
+
+func (s *statesInformer) gpuHealCheck(stopCh <-chan struct{}) {
+	count, ret := nvml.DeviceGetCount()
```
Comment (LambdaHJ) [inline pkg/koordlet/statesinformer/states_device.go:120]: Call nvml.Init first. It will return an error if driver not found.

Comment (LambdaHJ) [reply pkg/koordlet/statesinformer/states_device.go:120]: https://github.com/koordinator-sh/koordinator/blob/main/pkg/koordlet/metricsadvisor/collector_gpu_linux.go#L57

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 24 above)
Comment (buptcozy) [reply pkg/koordlet/statesinformer/states_gpu.go]: I mean this : https://github.com/NVIDIA/k8s-device-plugin/blob/master/internal/rm/health.go

File: pkg/koordlet/statesinformer/states_gpu.go (hunk 23 above)
Comment (buptcozy) [reply pkg/koordlet/statesinformer/states_gpu.go]: https://github.com/NVIDIA/k8s-device-plugin/blob/master/internal/rm/health.go can't change unhealthy to healthy too... so just follow it.

File: pkg/koordlet/statesinformer/states_device.go [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+	}
+	defer eventSet.Free()
+
+	for _, d := range devs {
+		device, ret := nvml.DeviceGetHandleByUUID(d)
+		if ret != nvml.SUCCESS {
+			klog.Errorf("failed to get device %s, err: %v", d, nvml.ErrorString(ret))
+			continue
+		}
+		ret = nvml.DeviceRegisterEvents(device, nvml.EventTypeXidCriticalError, eventSet)
```
Comment (buptcozy) [inline pkg/koordlet/statesinformer/states_device.go:181]: the whole health logic is little bit different from: https://github.com/NVIDIA/k8s-device-plugin/blob/master/internal/rm/health.go,   please make sure the diff is necessory.

File: cmd/koord-manager/main.go [hunk 29]
```diff
@@ -165,11 +169,13 @@ func main() {
...
-		SyncContext: noderesource.NewSyncContext(),
-		Clock:       clock.RealClock{},
-		Recorder:    mgr.GetEventRecorderFor("noderesource-controller"),
+		Recorder:       mgr.GetEventRecorderFor("noderesource-controller"),
+		Client:         mgr.GetClient(),
+		Scheme:         mgr.GetScheme(),
+		BESyncContext:  noderesource.NewSyncContext(),
+		GPUSyncContext: noderesource.NewSyncContext(),
+		Clock:          clock.RealClock{},
+		DeviceClient:   v1alpha1.NewForConfigOrDie(mgr.GetConfig()).Devices(),
```
Comment (eahydra) [inline cmd/koord-manager/main.go:178]: 需要创建这个DeviceClient么。controller-runtime可以直接CRUD CRD的呀。

File: pkg/slo-controller/noderesource/noderesource.go [hunk 30]
```diff
@@ -115,6 +116,85 @@ func (r *NodeResourceReconciler) resetNodeBEResource(node *corev1.Node, reason,
...
+func (r *NodeResourceReconciler) isGPUResourceNeedSync(new, old *corev1.Node) bool {
+	strategy := config.GetNodeColocationStrategy(r.cfgCache.GetCfgCopy(), new)
+
+	lastUpdatedTime, ok := r.GPUSyncContext.Load(util.GenerateNodeKey(&new.ObjectMeta))
+	if !ok || r.Clock.Now().After(lastUpdatedTime.Add(time.Duration(*strategy.UpdateTimeThresholdSeconds)*time.Second)) {
+		klog.Warningf("node %v resource expired, need sync", new.Name)
+		return true
+	}
+
+	if util.IsResourceDiff(old.Status.Allocatable, new.Status.Allocatable, extension.GPUMemoryRatio, *strategy.ResourceDiffThreshold) {
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/noderesource.go:128]: 建议考虑换个实现...

```go
for _, resourceName := range []corev1.ResourceName{extension.GPUMemoryRatio, extension.GPUMemoryRatio, extension.GPUMemory} {
   if util.IsResourceDiff(old.Status.Allocatable, new.Status.Allocatable, resourceName, *strategy.ResourceDiffThreshold) {
      return true
   }
}
return false
```

File: pkg/slo-controller/noderesource/noderesource.go [hunk 31]
```diff
@@ -115,6 +116,85 @@ func (r *NodeResourceReconciler) resetNodeBEResource(node *corev1.Node, reason,
...
+	copyNode := node.DeepCopy()
+	copyNode.Status.Allocatable[extension.GPUCore] = *coreTotal
+	copyNode.Status.Allocatable[extension.GPUMemory] = *memoryTotal
+	copyNode.Status.Allocatable[extension.GPUMemoryRatio] = *ratioTotal
+
+	if !r.isGPUResourceNeedSync(copyNode, node) {
+		return nil
+	}
+
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Comment (eahydra) [inline pkg/slo-controller/noderesource/noderesource.go:169]: 还有TooManyRequests也兼容下吧。

File: pkg/koordlet/statesinformer/states_device.go (hunk 28 above)
Comment (eahydra) [reply pkg/koordlet/statesinformer/states_device.go:181]: +1

Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_device.go:181]: updated

File: cmd/koord-manager/main.go (hunk 29 above)
Comment (ZYecho) [reply cmd/koord-manager/main.go:178]: I prefer use the crd client which generated by crd api

File: pkg/slo-controller/noderesource/noderesource.go (hunk 30 above)
Comment (ZYecho) [reply pkg/slo-controller/noderesource/noderesource.go:128]: nice, thanks

File: pkg/slo-controller/noderesource/noderesource.go (hunk 31 above)
Comment (ZYecho) [reply pkg/slo-controller/noderesource/noderesource.go]: done

File: cmd/koord-manager/main.go (hunk 29 above)
Comment (eahydra) [reply cmd/koord-manager/main.go:178]: 主要是没有必要吧. 生成出来的 client 又没有特殊的逻辑，整个组件又都是基于controller-runtime实现的，还是建议保持一致。

**Review by eahydra — COMMENTED** [review]
/lgtm

**Comment by jasonliu747** [comment]
Please also create a pull request to update RBAC accordingly in this repo https://github.com/koordinator-sh/charts. Thanks.

**Review by LambdaHJ — APPROVED** [review]
/lgtm

File: pkg/koordlet/statesinformer/states_informer.go [hunk 32]
```diff
@@ -170,6 +178,13 @@ func (s *statesInformer) Run(stopCh <-chan struct{}) error {
 	go s.startCallbackRunners(stopCh)
 
 	go wait.Until(s.reportNodeTopology, s.config.NodeTopologySyncInterval, stopCh)
+	if features.DefaultKoordletFeatureGate.Enabled(features.Accelerators) {
+		go wait.Until(s.reportDevice, s.config.NodeTopologySyncInterval, stopCh)
+		// check is nvml is available
+		if s.initGPU() {
```
Comment (hormes) [inline pkg/koordlet/statesinformer/states_informer.go:184]: If init fails, do you still need to start the reportDevice goroutine?

Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_informer.go:184]: node may contains fpga, but not contains gpu

File: pkg/koordlet/statesinformer/states_device.go [hunk 33]
```diff
@@ -0,0 +1,232 @@
...
+					Controller:         &blocker,
+					BlockOwnerDeletion: &blocker,
+				},
+			},
+		},
+		Spec: schedulingv1alpha1.DeviceSpec{
+			Devices: gpuDevices,
+		},
+	}
+	// TODO: compare diff before update
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/states_device.go:70]: please create an issue and add the link here, maybe also include other related crds like `NodeTopo`?

File: pkg/slo-controller/noderesource/noderesource_controller.go [hunk 34]
```diff
@@ -110,6 +112,18 @@ func (r *NodeResourceReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
 	}
+	device := &schedulingv1alpha1.Device{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: node.Name, Namespace: node.Namespace}, device)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("failed to get device %s, err: %v", node.Name, err)
+		}
+		return ctrl.Result{Requeue: false}, nil
+	}
+	if err := r.updateGPUNodeResource(node, device); err != nil {
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply pkg/slo-controller/noderesource/noderesource_controller.go:123]: could you provide some use case?

File: pkg/slo-controller/noderesource/noderesource_controller.go [hunk 35]
```diff
@@ -110,6 +112,18 @@ func (r *NodeResourceReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		klog.Errorf("failed to update node %v BE resource, error: %v", node.Name, err)
 		return ctrl.Result{Requeue: true}, err
 	}
+	device := &schedulingv1alpha1.Device{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: node.Name, Namespace: node.Namespace}, device)
+	if err != nil {
+		if !errors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply pkg/slo-controller/noderesource/noderesource_controller.go:118]: done

File: pkg/koordlet/statesinformer/states_device.go [hunk 36]
```diff
@@ -0,0 +1,232 @@
...
+	return &metriccache.QueryParam{
+		Aggregate: metriccache.AggregationTypeLast,
+		Start:     &start,
+		End:       &end,
+	}
+}
+
+func (s *statesInformer) reportDevice() {
+	s.nodeRWMutex.RLock()
+	copyNode := s.node.DeepCopy()
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply pkg/koordlet/statesinformer/states_device.go]: done

File: pkg/slo-controller/noderesource/noderesource_controller.go (hunk 34 above)
[the remark this answers is not in the document]
Comment (zwzhang0107) [reply pkg/slo-controller/noderesource/noderesource_controller.go:124]: if user only want to run koordinator on a cluster which doesnot have GPU, a `gpu=0` field on node is confusing.

**Pull request merged**